### PR TITLE
Recommend plain text output in addition to protobuf

### PR DESCRIPTION
**Pull request by barbeau**
**Best Practice Proposal:**

Feeds should provide a plain text output in addition to protocol buffer encoded content for debugging purposes. 

**Use Cases:** 

Often agencies and other non-technical staff want to check the raw contents of a feed (e.g., to see if an alert is still published). 

If the only output is in protocol buffer format, then the agency must use special tools (e.g., https://github.com/CUTR-at-USF/gtfs-realtime-validator) to decode that output and view a plain text version of the feed.

If the feed offers a plain text output (e.g., by simply appending `?debug` to the feed URL in a web browser) it increases transparency of the feed contents, making debugging and communication surrounding the feed contents more transparent.

**Data Examples:**

N/A

**Public-facing Implementation (Optional):**

* HART in Tampa - http://realtime.prod.obahart.org:8088/trip-updates?debug

### CODE REVIEW

**Comment by e-lo** [comment]
I like the general idea here.  Two suggestions:

1. Suggest recommending **"structured data that can be represented in UTF8** (i.e. JSON) rather than any sort of plain text
2. Recommend being explicit about format in the API call.  Rather than `debug`, suggest the approach used by [transit.511.org](https://511.org/open-data/transit) to have a passed parameter in the URI for `format=json`

**Comment by paulswartz** [comment]
+1 to JSON, rather than the plain-text Protobuf output.

MBTA provides a JSON version of our feeds next to our Protobuf ones: https://cdn.mbta.com/realtime/VehiclePositions.pb and https://cdn.mbta.com/realtime/VehiclePositions.json for example.
For reducing costs, these files are statically hosted, so we would need to re-architect our hosting in order to support using a query string to distinguish the two feeds.

**Comment by barbeau** [comment]
Good idea, I agree that being more specific than "plain text" here is better.

I'm hesitant to specify JSON as the only option here though, because IIRC the protocol buffer compiler tools from Google support a `.toString()` method that allows you to convert a message to plain text in a single line of code, and while the text looks like JSON it technically isn't valid JSON. Tampa's feed is one example and I've seen the same elsewhere:

```
header {
  gtfs_realtime_version: "1.0"
  incrementality: FULL_DATASET
  timestamp: 1642694516
}
entity {
  id: "trip_update_1732804"
  trip_update {
    trip {
      trip_id: "1732804"
    }
    ...
  }
}
```

So converting to valid JSON would entail additional development work, and I don't think that it's really necessary to have a valid JSON version of the feed for the human-readable use case here.

Here's a new first sentence that tries to address this, and also @paulswartz's good point about some files being statically hosted:

>Feeds should provide an option to display data in a structured UTF-8 text output (e.g., JSON) in addition to the protocol-buffer encoded feed content (e.g., by appending `&debug` to the feed URL, providing an alternate URL to different file such as `trip-updates.json`)...

**Comment by e-lo** [comment]
> I'm hesitant to specify JSON as the only option

Agree. You could extend my suggestion to just being more specific about "whatever format" is used ;-)

**Comment by derhuerst** [comment]
Probably this design is not deemed pragmatic enough, but keep in mind that [HTTP content negotiation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation) has been designed to support this use case:

- Returning different encodings of the same resource, based on the clients preferences.
- Given that (all AFAIK) browsers send `text/html`, falling back to `*/*`, the "automatic debug format" would already appear whenever someone opens the feed in the browser.
- It would be great to designate a mime type for GTFS-RT encoded via Protocol Buffers (e.g. [`application/x-protobuf;proto=com.google.transit.realtime`](https://stackoverflow.com/questions/30505408/what-is-the-correct-protobuf-content-type)).
- In order not to break clients that don't send a set of preferred encodings, we could recommend that whenever there's no such set, or whenever `text/plain`/`application/json` don't appear before `*/*`, the server should respond with pbf.

**Comment by paulswartz** [comment]
> Here's a new first sentence that tries to address this, and also @paulswartz's good point about some files being statically hosted:
> 
> > Feeds should provide an option to display data in a structured UTF-8 text output (e.g., JSON) in addition to the protocol-buffer encoded feed content (e.g., by appending `&debug` to the feed URL, providing an alternate URL to different file such as `trip-updates.json`)...

👍 You could call out the Protobuf textual format as well.

@derhuerst you're right that HTTP content negotiation would work for this, but it has the same issues that `&debug` does, where static hosting (Amazon S3 for example) doesn't support it.

**Comment by barbeau** [comment]
@derhuerst I agree, HTTP content negotiation is probably the "right" technical solution if we were deploying the system end-to-end ourselves, but given the mix of entities deploying solutions, as you say it's not pragmatic.

>You could call out the Protobuf textual format as well.

Good idea, I just pushed some edits including this.

**Comment by derhuerst** [comment]
> you're right that HTTP content negotiation would work for this, but it has the same issues that `&debug` does, where static hosting (Amazon S3 for example) doesn't support it.

I'd argue that the feature we're discussing here will be opt-in anyways, thus you can expect more technical tools/skills than with the mandatory part of GTFS-RT or even GTFS Static.

Although I have only seen a handful of GTFS-RT feeds out there (you probably have a better perspective), I haven't seen one where the skills or tools of the people in charge of implementing the feed generation we limited to plain AWS S3. Isn't implementing a GTFS-RT feed a highly technical feat *anyways*, due to the highly dynamic nature of realtime transit data, and given that Protocol Buffers are more complex than plain text?

AFAIK (not sure) all major CDNs (CloudFront, Cloudflare, fastly) can be configured to support content negotiation or do it out-of-the-box, depending on the use case.

---

I don't want to argue for something that we won't agree on though. If we decide to pick the pragmatic route (different URLs for different formats/encodings), what do you think about clearly denoting that the textual format is "derived" from (or follows the semantics of) Protocol Buffers, e.g. `.pbf.txt` or `.pbf.json`?

**Comment by barbeau** [comment]
>If we decide to pick the pragmatic route (different URLs for different formats/encodings), what do you think about clearly denoting that the textual format is "derived" from (or follows the semantics of) Protocol Buffers, e.g. .pbf.txt or .pbf.json?

@derhuerst Are you referring to the name of the file/URL path, or the contents of the file itself too?

I'm open to proposed wording if you have some - see the diff in the PR for the latest proposal. It's tricky to come up with something that's sufficiently descriptive without being overly prescriptive.

**Comment by derhuerst** [comment]
> > If we decide to pick the pragmatic route (different URLs for different formats/encodings), what do you think about clearly denoting that the textual format is "derived" from (or follows the semantics of) Protocol Buffers, e.g. .pbf.txt or .pbf.json?
> 
> Are you referring to the name of the file/URL path, or the contents of the file itself too?

Just about the file/URL path.

We could of course also add a note to the plain text version, but not to the JSON version, as JSON doesn't have a comment syntax.

> I'm open to proposed wording if you have some - see the diff in the PR for the latest proposal. It's tricky to come up with something that's sufficiently descriptive without being overly prescriptive.

I will try to come up with a more precise phrasing and propose it. 👍

**Comment by derhuerst** [comment]
The point of these recommendations is to *narrow down* the nr of different spec-compliant ways how to serve a GTFS-RT feed to one (or a few) specific way, right? They provide a "signal" to the community what is expected from a well-built and well-maintained GTFS-RT feed?

I've thought about this, and I think we should **recommend *one* textual format**, e.g. the [Protobuf TextFormat](https://developers.google.com/protocol-buffers/docs/reference/java/com/google/protobuf/TextFormat), and make clear that serving more (textual) formats like JSON under different URLs and/or `Content-Type`s is fine.

Given the "signalling effect" of these recommendations, adding *any* of the phrasings discussed here will already lead to many feed providers adapting their feeds, so why not use the opportunity to encourage *more usefulness*: If all well-built & well-maintained feeds provided at least *one specific* format, consumers would be able to work more effectively.

File: en/introduction.md [hunk 1]
```diff
@@ -21,6 +21,7 @@ Recommended practices are organized into two primary sections
 * The server hosting GTFS Realtime data should be reliable and consistently return validly-formatted protobuf-encoded responses. Fewer than 1% of responses should be invalid (protobuf errors or fetching errors).
 * The web-server hosting GTFS Realtime data should be configured to correctly report the file modification date (see HTTP/1.1 - Request for Comments 2616, under Section 14.29) so consumers can leverage the `If-Modified-Since` HTTP header. This saves producers and consumers bandwidth by avoiding transferring feed contents that haven't changed.
 * Feeds should provide protocol buffer-encoded feed content by default when queried via an HTTP request at the given URL - consumers should not need to define special HTTP accept headers to receive protocol-buffer encoded content.
```
Comment (derhuerst) [inline en/introduction.md:23]: I think we should also adapt this statement to only refer to HTTP requests *without* a matching `Accept` header. Those clients that send `text/plain` by default as one of the accepted MIME types would (and should!) receive the textual representation.

```suggestion
* Feeds should provide protocol buffer-encoded feed content by default when queried via an HTTP request without an `Accept` header. Consumers should not need to send an HTTP `Accept` header to receive protocol-buffer encoded content.
```

File: en/introduction.md [hunk 2]
```diff
@@ -21,6 +21,7 @@ Recommended practices are organized into two primary sections
...
+* Feeds should provide an option to display data in a structured UTF-8 text output (e.g., JSON, Protobuf [TextFormat](https://developers.google.com/protocol-buffers/docs/reference/java/com/google/protobuf/TextFormat)) in addition to the protocol-buffer encoded feed content (e.g., by appending `&debug` to the feed URL, providing an alternate URL to different file such as `trip-updates.json`). This enables a simple human-readable version of the information for troubleshooting purposes without requiring special tools to view the current contents of the real-time feed.
```
Comment (derhuerst) [inline en/introduction.md:24]:
```suggestion
* Feeds should provide an option to request their content in a structured UTF-8 text representation (e.g., JSON, [Protobuf TextFormat](https://developers.google.com/protocol-buffers/docs/reference/java/com/google/protobuf/TextFormat)) in addition to the protocol-buffer encoded feed content (e.g., by appending `&debug` to the feed URL, providing an alternate URL pathname such as `trip-updates.pb.json`). This enables opting into a simple human-readable version of the information for troubleshooting purposes, without requiring special tools to view the binary-encoded contents of the real-time feed. An appropriate `Content-Type` should be sent to denote which text representation is being served.
```

Comment (derhuerst) [reply en/introduction.md:24]: Bonus points for serving with >=1 `Link: <…>; rel="alternate"; type="…"` HTTP headers to let clients know where the alternative represenation(s) are!

related recommendations:
- https://www.w3.org/2001/tag/doc/alternatives-discovery.html
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Link
- https://datatracker.ietf.org/doc/html/rfc8288

**Comment by Sergiodero** [comment]
As GTFS Best Practices (BP) are currently in the process of being merged to the specification, MobilityData is migrating outstanding issues and PRs from this repository to google/transit. Thus, this PR will be closed and further discussion regarding this BP should take place in google/transit. Please refer to [Issue #421](https://github.com/google/transit/issues/421) for a more detailed explanation of the migration process and the proposed next steps. 

With this, we’re hoping to bring more visibility to outstanding BP issues and to restart the discussion around them, so that any improvement that the community finds valuable could be carried forward.

**Pull request closed without merging**